pkg/auth/dependency/webapp/service.go

```diff
@@ -0,0 +1,251 @@
+package webapp
+
+import (
+	"errors"
+	"net/url"
+
+	"github.com/authgear/authgear-server/pkg/auth/dependency/newinteraction"
+	"github.com/authgear/authgear-server/pkg/core/skyerr"
+)
+
+//go:generate mockgen -source=service.go -destination=service_mock_test.go -package webapp
+
+var ErrRollback = errors.New("rollback")
+
+type ErrorRedirectURIGetter interface {
+	GetErrorRedirectURI() string
+}
+
+type Store interface {
+	Create(state *State) error
+	Get(instanceID string) (*State, error)
+	DeleteFlow(flowID string) error
+}
+
+type GraphService interface {
+	Get(instanceID string) (*newinteraction.Graph, error)
+	WithContext(fn func(*newinteraction.Context) (*newinteraction.Graph, error)) error
+	NewGraph(ctx *newinteraction.Context, intent newinteraction.Intent) (*newinteraction.Graph, error)
+}
+
+type Service struct {
+	Store Store
+	Graph GraphService
+}
+
+func (s *Service) GetIntent(intent *Intent, stateID string) (state *State, graph *newinteraction.Graph, edges []newinteraction.Edge, err error) {
+	if stateID != "" {
+		return s.Get(stateID)
+	}
+
+	errRollback := s.Graph.WithContext(func(ctx *newinteraction.Context) (_ *newinteraction.Graph, err error) {
+		graph, err = s.Graph.NewGraph(ctx, intent.Intent)
+		if err != nil {
+			return
+		}
+
+		node := graph.CurrentNode()
+		edges, err = node.DeriveEdges(ctx, graph)
+		if err != nil {
+			return
+		}
+
+		err = ErrRollback
+		return
+	})
+	if !errors.Is(errRollback, ErrRollback) {
+		err = errRollback
+		return
+	}
+
+	state = &State{
+		FlowID:          newFlowID(),
+		InstanceID:      newInstanceID(),
+		RedirectURI:     intent.RedirectURI,
+		KeepState:       intent.KeepState,
+		GraphInstanceID: graph.InstanceID,
+	}
+
+	return
+}
+
+func (s *Service) Get(stateID string) (state *State, graph *newinteraction.Graph, edges []newinteraction.Edge, err error) {
+	state, err = s.Store.Get(stateID)
+	if err != nil {
+		return
+	}
+
+	graph, err = s.Graph.Get(state.GraphInstanceID)
+	if err != nil {
+		return
+	}
+
+	errRollback := s.Graph.WithContext(func(ctx *newinteraction.Context) (_ *newinteraction.Graph, err error) {
+		node := graph.CurrentNode()
+		edges, err = node.DeriveEdges(ctx, graph)
+		if err != nil {
+			return
+		}
+
+		err = ErrRollback
+		return
+	})
+	if !errors.Is(errRollback, ErrRollback) {
+		err = errRollback
+		return
+	}
+
+	return
+}
+
+func (s *Service) PostIntent(intent *Intent, inputer func() (interface{}, error)) (result *Result, err error) {
+	state := &State{
+		FlowID:      newFlowID(),
+		InstanceID:  newInstanceID(),
+		RedirectURI: intent.RedirectURI,
+		KeepState:   intent.KeepState,
+	}
+
+	var graph *newinteraction.Graph
+	var edges []newinteraction.Edge
+	// FIXME(webapp): How to convey the intention that we only want to
```

Comment:
In commit 4b62c07cb9f2914b20159df21d38852060a77dbf, I've fixed `WithContext` to handle the returned values better:
- nil graph, nil error -> commit without creating new graph instance, use at the end of graph.
- nil graph, non-nil error -> rollback without creating new graph instance, use when `Accept` returns a non-recognized error. the original graph should still be used.

Is that okay?
- non-nil graph, nil error -> commit AND create new graph instance, should not happen.
- non-nil graph, non-nil error -> rollback and creating new graph instance, use when `Accept`
returns `ErrRequireInput`.

Reply:
In WebAppService's GetXXX, commit is never expected.

In PostXXX, commit is still rare because it only happens when the graph ends. Maybe `fn` should return one more value to tell whether commit is desired.

The commit condition should be `err == nil && commit == true`
The create-graph condition should be `graph != nil`
The rollback condition is the negation of the commit condition
WithContext must either rollback or commit.